intermediate_source/parametrizations.py

```diff
@@ -0,0 +1,353 @@
+
+Regularizing deep-learning models is a surprisingly challenging task.
+Classical techniques such as penalty methods often fall short when applied
+on deep models due to the complexity of the function being optimized.
+This is particularly problematic when working with ill-conditioned models.
+Examples of these are RNNs trained on long sequences and GANs. A number
+of techniques have been proposed in the recent years to regularize these
+models and improve their convergence. On recurrent models, it has been
+proposed to control the singular values of the recurrent kernel for the
+RNN to be well-conditioned. This can be achieved, for example, by making
+the recurrent kernel `orthogonal <https://en.wikipedia.org/wiki/Orthogonal_matrix>`_.
+Another way to regularize recurrent models is via
+"`weight normalization <https://pytorch.org/docs/stable/generated/torch.nn.utils.weight_norm.html>`_".
+This approach proposes to decouple the learning of the parameters from the
+learning of their scale.  To do so, the parameter is divided by its
+`Frobenius norm <https://en.wikipedia.org/wiki/Matrix_norm#Frobenius_norm>`_.
+A similar regularization was proposed for GANs under the name of
+"`spectral normalization <https://pytorch.org/docs/stable/generated/torch.nn.utils.spectral_norm.html>`_". This method
+controls the Lipschitz constant of the network by dividing its parameters by
+their `spectral norm <https://en.wikipedia.org/wiki/Matrix_norm#Special_cases>`_,
+rather than its Frobenius norm.
+
+All these methods have a pattern in common. They all transform a parameter
+in an appropriate way before using it. In the first case, they make it orthogonal by
+using a function that maps matrices to orthogonal matrices. In the case of weight
+and spectral normalization, they divide the original parameter by its norm.
+
+More generally, all these examples use a function to put extra structure on the parameters.
+In other words, they use a function to constrain the parameters.
+
+In this tutorial, you will learn how to implement and use this pattern to write and
+put constraints on your model. Doing so is as easy as writing your own ``nn.Module``.
+
+Requirements
+------------
+``"torch>=1.9.0a0+7aeee28"``
+
+Implementing Parametrizations by Hand
+-------------------------------------
+
+Assume that we want to have a square linear layer with symmetric weights, that is,
+with weights :math:`X` such that :math:`X = X^{\intercal}`. One way to do so is
+to copy the upper triangular part of the matrix into its lower triangular part
+"""
+
+import torch
+import torch.nn as nn
+import torch.nn.utils.parametrize as P
+
+def symmetric(X):
+    return X.triu() + X.triu(1).transpose(-1, -2)
+
+X = torch.rand(3, 3)
+A = symmetric(X)
+print(A)
+assert torch.allclose(A, A.T)
+
+###############################################################################
+# We can then use this idea to implement a linear layer with symmetric weights:
+class LinearSymmetric(nn.Module):
```

Comment:
Less of a straw-man (for this simple case) would be a module deriving from Linear and adding a `weight` property

Reply:
That one's tricky. If you think about it, it'd go as something like this but with a bit more flare:
```python
import torch.nn as nn

class MyLin(nn.Linear):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self._weight = self.weight  # Why does this line work?
        delattr(self, "weight")

    @property
    def weight(self):
        return self._weight

lin = MyLin(3, 4)
```
The line above works because of how `nn.Module` handles `__getitem__` and `__setitem__`. One could almost say that works "out of pure chance". What happens in that line is:
- It calls `__getattribute__`
- `__getattribute__` finds the property `self.weight` and calls it
- The property looks for `self._weight`. At that time it does not exist, so it raises an `AttributeError`.
- Since `__getattribute__` got an `AttributeError`,  `nn.Module.__getattr__` is called
- `nn.Module.__getattr__` finds the `self.weight` that was created in `nn.Linear.__init__` and returns it

This is quite a mess really. That's why I went for the simpler method here.

Reply:
I had used `register_parameter('weight', None)` instead: https://gist.github.com/vadimkantorov/4f34fe60d2ef00e72dcad16512d224af

Reply:
Ah, the trick is that you still want to access old `_weight`, while I didn't need it

Reply:
That works for exactly the same convoluted reason, but changing `__getattr__` and `__getattribute__` with `__setattr__`  and `__setattribute__`. Again, it works, but it's tricky to know why it works.